// _vue-backgrounds@1.0.78@vue-backgrounds/src/engines/three/presents/fighters.js
export const fighters = {

  SpectraJ: {
    path: "/models/starpack/ship7.dae",
    scale: { x: 1, y: 1, z: -1 },
    front_cam: { y: -0.2, z: -3.3 },
    game_play_start_cam: { y: 1.1, z: 2 },
    game_play_ultimate_cam: { y: 1.1, z: 35.4 },
    preview_cam: { y: 7.6, z: 8.5 },
    first_person: { y: 0.5, z: 0 },
    top_view: { y: 25.2, z: 5.2 },
    maxSpeed: 2.1
  },
  Sx3c: {
    path: "/models/starpack/ship5.dae",
    scale: { x: 1, y: 1, z: -1 },
    front_cam: { y: -0.2, z: -3.3 },
    game_play_start_cam: { y: 1.1, z: 2 },
    game_play_ultimate_cam: { y: 1.1, z: 35.4 },
    preview_cam: { y: 7.6, z: 8.5 },
    first_person: { y: 0.5, z: 0 },
    top_view: { y: 25.2, z: 5.2 },
    maxSpeed: 2.1
  },
  Sx5c: {
    path: "/models/starpack/ship5.dae",
    scale: { x: 1, y: 1, z: -1 },
    front_cam: { y: -0.2, z: -3.3 },
    game_play_start_cam: { y: 1.1, z: 2 },
    game_play_ultimate_cam: { y: 1.1, z: 35.4 },
    preview_cam: { y: 7.6, z: 8.5 },
    first_person: { y: 0.5, z: 0 },
    top_view: { y: 25.2, z: 5.2 },
    maxSpeed: 2.5
  },
  SpectraR3: {
    path: "/models/starpack/ship8.dae",
    scale: { x: 1, y: 1, z: -1 },
    front_cam: { y: -0.2, z: -3.3 },
    game_play_start_cam: { y: 1.1, z: 2 },
    game_play_ultimate_cam: { y: 1.1, z: 35.4 },
    preview_cam: { y: 7.6, z: 8.5 },
    first_person: { y: 0.5, z: 0 },
    top_view: { y: 25.2, z: 5.2 },
    maxSpeed: 3.1
  },
  SpectraR4: {
    path: "/models/starpack/ship9.dae",
    scale: { x: 1, y: 1, z: -1 },
    front_cam: { y: -0.2, z: -3.3 },
    game_play_start_cam: { y: 1.1, z: 2 },
    game_play_ultimate_cam: { y: 1.1, z: 35.4 },
    preview_cam: { y: 7.6, z: 8.5 },
    first_person: { y: 0.5, z: 0 },
    top_view: { y: 25.2, z: 5.2 },
    maxSpeed: 2.1
  },
  SpectraTitan: {
    path: "/models/starpack/ship10.dae",
    scale: { x: 1, y: 1, z: -1 },
    front_cam: { y: 1.8, z: -4.5 },
    game_play_start_cam: { y: 1.1, z: 7.4 },
    game_play_ultimate_cam: { y: 1.1, z: 29.4 },
    preview_cam: { y: 8, z: 8 },
    first_person: { y: 0.5, z: 2.0 },
    top_view: { y: 48.6, z: 1.2 },
    maxSpeed: 1.1
  },

  Bami: {
    path: "/models/starpack/bemi/sk_389211.dae",
    scale: { x: 1, y: 1, z: -1 },
    front_cam: { y: 1.8, z: -4.5 },
    game_play_start_cam: { y: 1.1, z: 7.4 },
    game_play_ultimate_cam: { y: 1.1, z: 29.4 },
    preview_cam: { y: 8, z: 8 },
    first_person: { y: 0.5, z: 2.0 },
    top_view: { y: 48.6, z: 1.2 },
    maxSpeed: 1.1
  },

  Tempest: {
    path: "/models/starpack/tempest_a.dae",
    scale: { x: 1, y: 1, z: -1 },
    front_cam: { y: 1.8, z: -4.5 },
    game_play_start_cam: { y: 1.1, z: 7.4 },
    game_play_ultimate_cam: { y: 1.1, z: 29.4 },
    preview_cam: { y: 8, z: 8 },
    first_person: { y: 0.5, z: 2.0 },
    top_view: { y: 48.6, z: 1.2 },
    maxSpeed: 1.1
  }

}
