// _vue-backgrounds@1.0.78@vue-backgrounds/src/engines/pixi/ScaledContainer.js
import { Container, Point } from "pixi.js";
export default class ScaledContainer extends Container {
    constructor(vue) {
        super();
        this.currentSize = {
            w: 0,
            h: 0
        };
        const { width, height, canvasWidth, canvasHeight } = vue.$store.state.pixiUtil.Renderer;
        this.resizeHandler(window.innerWidth, window.innerHeight, canvasWidth, canvasHeight);
    }
    resizeHandler(rw, rh, tw = 1920, th = 1080) {
        const Xratio = rw / tw;
        const Yratio = rh / th;
        let scaleRatio = rw > rh ? Xratio : Yratio;
        let scale = new Point(scaleRatio, scaleRatio);
        let offsetX = rw / 2 - tw * scaleRatio / 2;
        let offsetY = rh / 2 - th * scaleRatio / 2;
        if (th * scaleRatio < rh) {
            scaleRatio = Yratio;
            scale = new Point(scaleRatio, scaleRatio);
            offsetX = rw / 2 - tw * scaleRatio / 2;
            offsetY = rh / 2 - th * scaleRatio / 2;
        }
        this.position.x = offsetX;
        this.position.y = offsetY;
        this.scale = scale;
    }
}
