// _vue-backgrounds@1.0.78@vue-backgrounds/src/pixicom/Sprite.js
import { Sprite, BLEND_MODES, Texture } from "pixi.js"
import Container from "./Container.js"

export default {
  mixins: [Container],
  props: {
    anchorX: {
      type: Number,
      required: false,
      default: 0
    },
    anchorY: {
      type: Number,
      required: false,
      default: 0
    },
    blendMode: {
      type: BLEND_MODES,
      required: false,
      default: BLEND_MODES.NONE
    },
    // buttonMode
    // pluginName
    // shader
    // texture
    tint: {
      type: Number,
      required: false,
      default: 0xFFFFFF
    },
    texture: {
      type: Texture,
      required: true,
      default: Texture.EMPTY
    }
  },
  computed: {
    instance() {
      return new Sprite(this.texture)
    }
  },
  watch: {
    instance: {
      handler(instance) {
        if (this.tint) {
          instance.text = this.tint
        }
        if (this.blendMode) {
          instance.blendMode = this.blendMode
        }
        if (this.anchorX || this.anchorY) {
          instance.anchor.set(this.anchorX || 0, this.anchorY || 0)
        }
      },
      immediate: true
    },
    tint(tint) {
      this.instance.tint = tint
    },
    blendMode(blendMode) {
      this.instance.blendMode = blendMode
    },
    anchorX(anchorX) {
      this.instance.anchor.x = anchorX
    },
    anchorY(anchorY) {
      this.instance.anchor.y = anchorY
    }
  }
}
